### DISCUSSION TITLE
videos autoselect: add layout choosing table in setup.toml file

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#25] [Ideas]
Based on https://discourse.coreelec.org/t/graphical-front-panel-display/12932/192

one could end up with many layouts that are not needed, one solution could be to add a table in the `setup.toml` file that specify which layout to use 

It will avoid creating identical layout.

Eg in my case:

```
PVR--> pvr layout (but some users could want a TV layout used for PVR and live TV)
live TV--> live tv layout (but some users could want a TV layout used for PVR and live TV)
TVshows--> tv show layout
movie--> movie layout (but some users could want a "other" layout)
addon--> default layout (I don't plan to use it)
music video--> default layout (I don't plan to use it)
other -->default layout
```
Of course all the used layout would have to be checked to make sure they exists

**Comment by mattblovell** [comment]
I'm not sure I presently know a heuristic for all of the choices listed above.  How does one distinguish a music video, for instance?

**Comment by nico1080** [comment]
My suggestion was not about the heuristic to detect witch layout to use. But on a setting that will user specify what layout to use based on the heuristic result,

Lets take the exemple of an user that want a specific layout for TV show and the defaut layout for all other case.
Curently he will need to create a TV layout and default layout (normal) and then copy/paste the default layout to PVR/LiveTV/ and movie layout Then if some modifications need to be done in the default layout it will need to be copy/pasted again.

With my suggestion he still  will need to create a TV layout and default layout (normal) and then specify in the `setup`.toml file something like
```
PVR--> use default layout
live TV--> use default layout
TVshows--> use tv show layout
movie--> use default layout
other --> use default layout
```
It could also allow for user custom name for each layout (eg a user could create a 'TV' layout to be used for PVR and liveTV sections)
Also if a layout is not declared (missing), it could failback to the default layout.

Heuristic for addon and music video could be implemented, but it is an other subject. (I will open a discusion on that)

**Reply by mattblovell** [reply]
> My suggestion was not about the heuristic to detect witch layout to use. But on a setting that will user specify what layout to use based on the heuristic result

Defining lookup table entries that code cannot (or doesn't know how to) reference isn't too useful, though. :) 

While the layout name independence will be nice, I fear that the ultimate direction for auto-selection would be to employ a user-specified heuristic.  That's not conveniently done in TOML.  If that is the direction, then we may as well go there, rather than adding this particular bit of infrastructure.

**Reply by nico1080** [reply]
Effectively a user based heuristic could be better than my suggestion. It could also allow user to define multiple layouts and use tehem based on any heuristric 

eg: if video have cover--> use layoutA else layout B
or if video lenght<20min --> use layout for short video else layout for long video

Or more applicable exemple to my setup in the PVR section: There could be some heuristic to detect if the recording played is finished or not and then use different layout.

**Comment by mattblovell** [comment]
I should note that, with recent changes to how auto-selection goes about checking the AUTOSELECT heuristic, a user only has to define the layouts that they wish to use.  For instance, one could have just `V_LIVETV` and `V_MOVIE` entries.

With v1.22, I modified `video_screens()` such that the heuristic makes an assignment to the `video_dmode` variable.  That then permits the name of the select layout (as a string) to be made available to any of the new callback functions (entries within the `ELEMENT_CB` or `STRING_CB` tables), should that be helpful.

The next step is to move that heuristic if-elif tree to a separate function, invoked through a variable.  The variable then provides a hook by which external code could replace the if-elif logic completely.

**Comment by mattblovell** [comment]
kodi_panel v1.24 moves the video-mode selection to a standalone function that gets passed the full video InfoLabel dictionary.  The invoking script can override the variable that points to that function, permitting for full customization of the heuristic.

I also copied this framework to apply to the audio-side of things.  So, should someone arrive at a means for deciding that there are distinct forms of audio file, that can guide audio layout selection.
